recipes/kfserving.rb

```diff
@@ -0,0 +1,208 @@
+  code <<-EOH
+    cd istio
+    istio_tar="istio-#{node['kube-hops']['istio_version']}-linux-amd64.tar.gz"
+    wget -nc #{node['kube-hops']['istio_url']}
+    tar zxf $istio_tar
+    rm -f $istio_tar
+    rm -f istio
+    ln -s istio-#{node['kube-hops']['istio_version']} istio
+    EOH
+  not_if { File.exist? "/home/#{node['kube-hops']['user']}/istio/istio-#{node['kube-hops']['istio_version']}/bin/istioctl" }
+end
+
+template "/home/#{node['kube-hops']['user']}/istio/istio-operator.yaml" do
+  source "istio-operator.yml.erb"
+  owner node['kube-hops']['user']
+  group node['kube-hops']['group']
+end
+
+bash 'apply-istio' do
+  user node['kube-hops']['user']
+  group node['kube-hops']['group']
+  environment ({ 'HOME' => ::Dir.home(node['kube-hops']['user']) })
+  cwd ::Dir.home(node['kube-hops']['user'])
+  code <<-EOH
+    cd istio
+    istio/bin/istioctl install -f istio-operator.yaml
+    EOH
+end
+
+# Knative
+
+directory "/home/#{node['kube-hops']['user']}/knative" do
+  owner node['kube-hops']['user']
+  group node['kube-hops']['group']
+  mode '0700'
+  action :create
+end
+
+template "/home/#{node['kube-hops']['user']}/knative/knative-serving-crds.yaml" do
+  source "knative-serving-crds.yml.erb"
+  owner node['kube-hops']['user']
+  group node['kube-hops']['group']
+end
+
+template "/home/#{node['kube-hops']['user']}/knative/knative-serving.yaml" do
+  source "knative-serving.yml.erb"
+  owner node['kube-hops']['user']
+  group node['kube-hops']['group']
+end
+
+template "/home/#{node['kube-hops']['user']}/knative/knative-istio.yaml" do
+  source "knative-istio.yml.erb"
+  owner node['kube-hops']['user']
+  group node['kube-hops']['group']
+end
+
+kube_hops_kubectl 'apply-knative-serving-crds' do
+  user node['kube-hops']['user']
+  group node['kube-hops']['group']
+  url "/home/#{node['kube-hops']['user']}/knative/knative-serving-crds.yaml"
+end
+
+kube_hops_kubectl 'apply-knative-serving' do
+  user node['kube-hops']['user']
+  group node['kube-hops']['group']
+  url "/home/#{node['kube-hops']['user']}/knative/knative-serving.yaml"
+end
+
+kube_hops_kubectl 'apply-knative-istio-controller' do
+  user node['kube-hops']['user']
+  group node['kube-hops']['group']
+  url "/home/#{node['kube-hops']['user']}/knative/knative-istio.yaml"
+end
+
+# Cert-manager
+
+template "/home/#{node['kube-hops']['user']}/cert-manager.yaml" do
+  source "cert-manager.yml.erb"
+  owner node['kube-hops']['user']
+  group node['kube-hops']['group']
+end
+
+kube_hops_kubectl 'apply-cert-manager' do
+  user node['kube-hops']['user']
+  group node['kube-hops']['group']
+  flags "--validate=false"
+  url "/home/#{node['kube-hops']['user']}/cert-manager.yaml"
+end
+
+bash 'wait-for-cert-manager' do
+  user node['kube-hops']['user']
+  group node['kube-hops']['group']
+  environment ({ 'HOME' => ::Dir.home(node['kube-hops']['user']) })
+  code <<-EOH
+    kubectl wait --for=condition=Ready pods --all -n cert-manager --timeout=60s
+    sleep 15s
```

Comment:
why this sleep here? isn't `timeout` above enough?

Reply:
No, It's not enough. The line above waits for the pod condition to be `Ready`, but it's not done with the installation/configuration. Without `sleep 15s` the installation of KFServing immediately after (line 124) fails since the cert-manager webhook is not ready yet.